io/src/main/scala/fs2/io/io.scala

```diff
@@ -28,19 +28,21 @@ package object io {
     * Reads all bytes from the specified `InputStream` with a buffer size of `chunkSize`.
     * Set `closeAfterUse` to false if the `InputStream` should not be closed after use.
     *
-    * This will block a thread in the `ExecutionContext`, so the size of any associated
-    * threadpool should be sized appropriately.
+    * Like `readInputStream` but each read operation is performed on the supplied execution
+    * context. Reads are blocking so the execution context should be configured appropriately.
     */
-  def readInputStreamAsync[F[_]](fis: F[InputStream],
-                                 chunkSize: Int,
-                                 closeAfterUse: Boolean = true)(
-      implicit F: Concurrent[F],
-      ec: ExecutionContext): Stream[F, Byte] = {
-    def readAsync(is: InputStream, buf: Array[Byte]) =
-      async.fork(readBytesFromInputStream(is, buf)).flatMap(_.join)
-
-    readInputStreamGeneric(fis, F.delay(new Array[Byte](chunkSize)), readAsync, closeAfterUse)
-  }
+  def readInputStreamAsync[F[_]](
+      fis: F[InputStream],
+      chunkSize: Int,
+      blockingExecutionContext: ExecutionContext,
+      closeAfterUse: Boolean = true)(implicit F: Async[F], timer: Timer[F]): Stream[F, Byte] =
+    readInputStreamGeneric(
+      fis,
+      F.delay(new Array[Byte](chunkSize)),
+      (is, buf) =>
+        Async.shift(blockingExecutionContext) *> readBytesFromInputStream(is, buf) <* timer.shift,
```

Comment:
@SystemFw Should this shift/shift-back pattern be done via `bracket`?

Reply:
I think it should actually, :+1: